pkg/ccl/testccl/sqlccl/sqlccl_test: TestReadCommittedVolatileUDF failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestReadCommittedVolatileUDF [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19725947?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19725947?buildTab=artifacts#/) on master @ [1f853825ea98c442d0a66c6d50bd1c1e644ad1d3](https://github.com/cockroachdb/cockroach/commits/1f853825ea98c442d0a66c6d50bd1c1e644ad1d3):


```
=== RUN   TestReadCommittedVolatileUDF
I250519 16:50:14.383681 1 (gostd) rand.go:256  [-] 1  random seed: 545304631242617657
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/5fb4f5d151a27410c45fc568213a2a4f/logTestReadCommittedVolatileUDF1506374610
    test_log_scope.go:76: use -show-logs to present logs inline
    read_committed_test.go:279: 
        	Error Trace:	pkg/ccl/testccl/sqlccl/read_committed_test.go:279
        	Error:      	Received unexpected error:
        	            	expected first invocation result to be 10; got 2
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | pkg/ccl/testccl/sqlccl/sqlccl_test_test.TestReadCommittedVolatileUDF.func2
        	            	  | 	pkg/ccl/testccl/sqlccl/read_committed_test.go:250
        	            	  | pkg/ccl/testccl/sqlccl/sqlccl_test_test.TestReadCommittedVolatileUDF.Group.GoCtx.func4
        	            	  | 	pkg/util/ctxgroup/ctxgroup.go:199
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1223
        	            	Wraps: (2) expected first invocation result to be 10; got 2
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestReadCommittedVolatileUDF
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5fb4f5d151a27410c45fc568213a2a4f/logTestReadCommittedVolatileUDF1506374610
--- FAIL: TestReadCommittedVolatileUDF (1.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedVolatileUDF.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50826

--- a/pkg/ccl/testccl/sqlccl/read_committed_test.go
+++ b/pkg/ccl/testccl/sqlccl/read_committed_test.go
@@ -8,6 +8,7 @@ package sqlccl_test
 import (
 	"context"
 	gosql "database/sql"
+	"fmt"
 	"strings"
 	"sync"
 	"sync/atomic"
@@ -19,7 +20,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/isolation"
 	"github.com/cockroachdb/cockroach/pkg/sql"
-	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
@@ -207,15 +207,42 @@ func TestReadCommittedVolatileUDF(t *testing.T) {
 
 	ctx := context.Background()
 	var readCommittedRetryCount atomic.Int64
-	params := base.TestServerArgs{}
-	params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
-		OnReadCommittedStmtRetry: func(retryReason error) {
-			// Track retries since we don't want them to happen in this test, since
-			// they would change the read timestamp.
-			readCommittedRetryCount.Add(1)
+	var trapRCTxn atomic.Bool
+	var tableID int
+	rcTxnBlocked, unblockRCTxn := make(chan struct{}, 1), make(chan struct{})
+	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			SQLExecutor: &sql.ExecutorTestingKnobs{
+				OnReadCommittedStmtRetry: func(retryReason error) {
+					// Track retries since we don't want them to happen in this
+					// test, since they would change the read timestamp.
+					readCommittedRetryCount.Add(1)
+				},
+			},
+			Store: &kvserver.StoreTestingKnobs{
+				TestingRequestFilter: func(_ context.Context, req *kvpb.BatchRequest) *kvpb.Error {
+					if !trapRCTxn.Load() {
+						return nil
+					}
+					if len(req.Requests) != 2 {
+						return nil
+					}
+					for _, r := range req.Requests {
+						get, ok := r.GetInner().(*kvpb.GetRequest)
+						if !ok {
+							return nil
+						}
+						if !strings.Contains(get.Key.String(), fmt.Sprintf("/Table/%d", tableID)) {
+							return nil
+						}
+					}
+					rcTxnBlocked <- struct{}{}
+					<-unblockRCTxn
+					return nil
+				},
+			},
 		},
-	}
-	s, sqlDB, _ := serverutils.StartServer(t, params)
+	})
 	defer s.Stopper().Stop(ctx)
 
 	_, err := sqlDB.Exec(`CREATE TABLE kv (k TEXT PRIMARY KEY, v INT) WITH (sql_stats_automatic_collection_enabled = false);`)
@@ -223,6 +250,8 @@ func TestReadCommittedVolatileUDF(t *testing.T) {
 	_, err = sqlDB.Exec(`INSERT INTO kv VALUES ('a', 10);`)
 	require.NoError(t, err)
 	_, err = sqlDB.Exec(`CREATE FUNCTION f() RETURNS INT AS $$ SELECT v FROM kv WHERE k = 'a' OR k = 'b' ORDER BY v LIMIT 1 FOR SHARE $$ LANGUAGE SQL VOLATILE;`)
+	require.NoError(t, err)
+	err = sqlDB.QueryRow("SELECT 'kv'::REGCLASS::OID").Scan(&tableID)
 	require.NoError(t, err)
 
 	g := ctxgroup.WithContext(ctx)
@@ -236,6 +265,7 @@ func TestReadCommittedVolatileUDF(t *testing.T) {
 
 	// Start a READ COMMITTED transaction that is blocked on txSerializable, and
 	// which tries to read a key that does not exist yet using a UDF.
+	trapRCTxn.Store(true)
 	txReadCommitted, err := sqlDB.BeginTx(ctx, &gosql.TxOptions{Isolation: gosql.LevelReadCommitted})
 	require.NoError(t, err)
 	var executedUDF atomic.Bool
@@ -259,18 +289,9 @@ func TestReadCommittedVolatileUDF(t *testing.T) {
 	// In a third transaction, add another row to the table after confirming that
 	// txReadCommitted is blocked. This row should not be visible to
 	// txReadCommitted's UDF.
-	testutils.SucceedsSoon(t, func() error {
-		var blockedCount int
-		if err := sqlDB.QueryRow(
-			`SELECT count(*) FROM crdb_internal.cluster_locks WHERE table_name = 'kv'`,
-		).Scan(&blockedCount); err != nil {
-			return err
-		}
-		if blockedCount == 0 {
-			return errors.Newf("expected to find blocked transaction")
-		}
-		return nil
-	})
+	<-rcTxnBlocked
+	trapRCTxn.Store(false)
+	close(unblockRCTxn)
 	_, err = sqlDB.Exec(`INSERT INTO kv VALUES ('b', 2)`)
 	require.NoError(t, err)
 

